webapp/tests/karma/ts/services/rules-engine.service.spec.ts

```diff
@@ -0,0 +1,619 @@
+      _id: 'taskdoc',
+      'emission.other': true,
+    });
+    expect(telemetryService.record.callCount).to.equal(3);
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 2]);
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:tasks:some-contacts');
+  });
+
+  it('fetchTargets() should send correct range to Rules Engine Core when getting targets', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    service = TestBed.inject(RulesEngineService);
+
+    const actual = await service.fetchTargets();
+
+    expect(actual).to.deep.eq([]);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.args[0][0]).to.have.keys('start', 'end');
+  });
+
+  it('should ensure freshness of tasks and targets', async () => {
+    rulesEngineCoreStubs.getDirtyContacts.returns(Array.from({ length: 20 }).map(i => i));
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    await service.isEnabled();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+    
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(3);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 20]);
+    expect(telemetryService.record.args[2]).to.deep.equal(['rules-engine:targets:dirty-contacts', 20]);
+  });
+
+  it('should ensure freshness of tasks only', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    await service.fetchTargets();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(5);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1][0]).to.equal('rules-engine:targets:dirty-contacts');
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTargetFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:targets');
+    expect(telemetryService.record.args[4][0]).to.equal('rules-engine:tasks:dirty-contacts');
+  });
+
+  it('should cancel all ensure freshness threads', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    fetchTasksResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    service.fetchTaskDocsForAllContacts();
```

Comment:
is there a reason you're not `await`-ing this? The original test called these two functions, in this sequence: 
```
 await service.fetchTargets();
 await service.fetchTaskDocsForAllContacts();
 ```

Reply:
If I use the sequence from the original test, this test fails. Is it possible something is not working correctly?

Reply:
Fixed